File: Python/Concurrent Algorithms/Semaphore.py

from enum import Enum
from queue import Queue

class Semaphore:
	class Value(Enum):
		print('Semaphore help in synchronization')
		Zero = 0
		One = 1

	def __init__(self):
		self.q = Queue()
		self.value = Semaphore.Value.One

	def P(self, s, p):
		if s.value == Semaphore.Value.One:
			s.value = Semaphore.Value.Zero
		else:
			# add the process to the waiting queue
			s.q.put(p)
			p.Sleep()

	def V(self, s):
		if s.q.qsize() == 0:
			s.value = Semaphore.Value.One
		else:
			# select a process from waiting queue
			p = s.q.queue[0]
			# remove the process from waiting as it has
			# been sent for CS
			s.q.get()
			p.Wakeup()

File: Java/Concurrent Algorithms/Semaphore.java
import java.util.*;

class Semaphore {
    System.out.println("Semaphore help in synchronization");
	public enum Value { Zero, One }
	public Queue<Process> q = new LinkedList<Process>();
	public Value value = Value.One;

	public void P(Semaphore s, Process p)
	{
		if (s.value == Value.One) {
			s.value = Value.Zero;
		}
		else {
			// add the process to the waiting queue
			q.add(p);
			p.Sleep();
		}
	}

	public void V(Semaphore s)
	{
		if (s.q.size() == 0) {
			s.value = Value.One;
		}
		else {
			// select a process from waiting queue
			Process p = q.peek();
			// remove the process from waiting as it has
			// been sent for CS
			q.remove();
			p.Wakeup();
		}
	}
}
